odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+            if out.dtype != self.values.dtype:
+                raise ValueError('output dtype {} not equal to expected '
+                                 'dtype {}'
+                                 ''.format(out.dtype, self.values.dtype))
+
+        indices, norm_distances = self._find_indices(x)
+        return self._evaluate(indices, norm_distances, out)
+
+    def _find_indices(self, x):
+        """Find indices and distances of the given nodes.
+
+        Can be overridden by subclasses to improve efficiency.
+        """
+        # find relevant edges between which xi are situated
+        index_vecs = []
+        # compute distance to lower edge in unity units
+        norm_distances = []
+
+        # iterate through dimensions
+        for xi, cvec in zip(x, self.coord_vecs):
+            idcs = np.searchsorted(cvec, xi) - 1
+
+            idcs[idcs < 0] = 0
+            idcs[idcs > cvec.size - 2] = cvec.size - 2
+            index_vecs.append(idcs)
+
+            norm_distances.append((xi - cvec[idcs]) /
+                                  (cvec[idcs + 1] - cvec[idcs]))
+
+        return index_vecs, norm_distances
+
+    def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluation method, needs to be overridden."""
+        raise NotImplementedError('abstract method')
+
+
+class _NearestInterpolator(_Interpolator):
+
+    """Nearest neighbor interpolator.
+
+    The code is adapted from SciPy's `RegularGridInterpolator
+    <http://docs.scipy.org/doc/scipy/reference/generated/\
+scipy.interpolate.RegularGridInterpolator.html>`_ class.
+
+    This implementation is faster than the more generic one in the
+    `_PerAxisPointwiseInterpolator`. Compared to the original code,
+    support of ``'left'`` and ``'right'`` variants are added.
+    """
+
+    def __init__(self, coord_vecs, values, input_type, variant):
+        """Initialize a new instance.
+
+        coord_vecs : sequence of `numpy.ndarray`'s
+            Coordinate vectors defining the interpolation grid
+        values : `array-like`
+            Grid values to use for interpolation
+        input_type : {'array', 'meshgrid'}
+            Type of expected input values in ``__call__``
+        variant : {'left', 'right'}
+            Indicates which neighbor to prefer in the interpolation
+        """
+        super(_NearestInterpolator, self).__init__(
+            coord_vecs, values, input_type)
+        variant_ = str(variant).lower()
+        if variant_ not in ('left', 'right'):
+            raise ValueError("variant '{}' not understood".format(variant_))
+        self.variant = variant_
+
+    def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluate nearest interpolation."""
+        idx_res = []
+        for i, yi in zip(indices, norm_distances):
+            if self.variant == 'left':
+                idx_res.append(np.where(yi <= .5, i, i + 1))
+            else:
```

Comment:
Specify that this is the `right` case, and that all others are rule out by above checks?

Reply:
Clarified